tool/planet/agent.go

```diff
@@ -108,151 +104,126 @@ func startLeaderClient(conf *LeaderConfig, errorC chan error) (leaderClient io.C
 	if conf.Role == RoleMaster {
-		switch conf.ElectionEnabled {
-		case true:
-			log.Infof("adding voter for IP %v", conf.PublicIP)
-			ctx, cancelVoter = context.WithCancel(context.TODO())
-			if err = client.AddVoter(ctx, conf.LeaderKey, conf.PublicIP, conf.Term); err != nil {
+		etcdapi := etcd.NewKeysAPI(etcdClient)
+		// Set initial value of election participation mode
+		_, err = etcdapi.Set(context.TODO(), conf.ElectionKey,
+			strconv.FormatBool(conf.ElectionEnabled),
+			&etcd.SetOptions{PrevExist: etcd.PrevNoExist})
+		if err != nil {
+			if err = convertError(err); !trace.IsAlreadyExists(err) {
 				return nil, trace.Wrap(err)
 			}
-		case false:
-			log.Info("shutting down services until election has been re-enabled")
-			// Shut down services at startup if running as master
-			if err := unitsCommand("stop"); err != nil {
-				log.Infof("failed to stop units: %v", err)
-			}
 		}
 	}
-	// watch the election mode status and start/stop participation
-	// depending on the value of the election mode key
-	client.AddWatchCallback(conf.ElectionKey, conf.Term, func(key, prevVal, newVal string) {
-		var err error
-		enabled, _ := strconv.ParseBool(newVal)
-		switch enabled {
-		case true:
-			if cancelVoter != nil {
-				log.Infof("voter is already active")
-				return
-			}
-			// start election participation
-			ctx, cancelVoter = context.WithCancel(context.TODO())
-			if err = client.AddVoter(ctx, conf.LeaderKey, conf.PublicIP, conf.Term); err != nil {
-				log.Errorf("failed to add voter for %v: %v", conf.PublicIP, trace.DebugReport(err))
-				errorC <- err
-			}
-		case false:
-			if cancelVoter == nil {
-				log.Info("no voter active")
-				return
-			}
-			// stop election participation
-			cancelVoter()
-			cancelVoter = nil
-		}
-	})
-	// modify /etc/hosts upon election of a new leader node
-	client.AddWatchCallback(conf.LeaderKey, conf.Term/3, func(key, prevVal, newLeaderIP string) {
-		if err := updateDNS(conf, hostname, newLeaderIP); err != nil {
-			log.Error(trace.DebugReport(err))
-		}
-	})
 
 	return client, nil
 }
 
 func writeLocalLeader(target string, masterIP string) error {
-	contents := fmt.Sprint(masterIP, " ",
+	contents := fmt.Sprintln(masterIP, " ",
 		constants.APIServerDNSName, " ",
 		constants.APIServerDNSNameGravity, " ",
 		constants.RegistryDNSName, " ",
-		LegacyAPIServerDNSName, "\n")
-	err := ioutil.WriteFile(
-		target,
-		[]byte(contents),
-		SharedFileMask,
-	)
+		LegacyAPIServerDNSName)
+	err := ioutil.WriteFile(target, []byte(contents), SharedFileMask)
 	return trace.Wrap(err)
 }
 
-func updateDNS(conf *LeaderConfig, hostname string, newMasterIP string) error {
-	log.Infof("Setting new leader address to %v in %v", newMasterIP, CoreDNSHosts)
-	err := writeLocalLeader(CoreDNSHosts, newMasterIP)
-	if err != nil {
-		return trace.Wrap(err)
+func updateControlPlaneUnits(localAddr string) handlerFunc {
+	return func(ctx context.Context, _, newLeaderAddr string, logger log.FieldLogger) error {
```

Comment:
Is the ignored (`_`) parameter here the old leader address? Why not use it as a logger field as well? Same below.

Reply:
Oh, missed that one - will add to the list of fields.